feat(ci): Add ASAN and UBSAN build and CI job

czentgr opened the pull request.

kgpai commented on .github/workflows/scheduled.yml:
```diff
@@ -1475,3 +1475,101 @@ jobs:
+            "-DVELOX_ENABLE_REMOTE_FUNCTIONS=ON"
+          )
+          clang++ --version
+          dnf install -y compiler-rt
```
nit: lets add this to the centos image

kgpai reviewed (commented):
Looks good , Thanks for this !

czentgr commented:
@kgpai I migrated the job to use the new steps for build/test. Repeated run is now 60mins for build and test. 
Please take another look. I think it is ready to go after review.

kgpai commented:
Thanks for putting this together — having ASAN/UBSAN coverage in CI is a great addition, and the integration with the existing build/test/status pattern is clean. I had a few questions and small suggestions on the changes; nothing blocking, just hoping to better understand the intent in a couple of places.

**1. `linux-build-base.yml` — `VELOX_BUILD_SHARED=ON` vs. the comment above it**

The comment says `VELOX_BUILD_SHARED is turned off due producing a SEGV if the shared library is used. Revisit once PR 16019 lands that removes thrift.` — but the flag list right below has `-DVELOX_BUILD_SHARED=ON`. Could you clarify which one is right? If the SEGV is still a real risk we probably want `OFF` here; if PR #16019 has landed and the comment is stale, would you mind updating the comment so the next reader doesn't have to wonder?

**2. `CMakeLists.txt` — `VERSION_GREATER_EQUAL 15` → `VERSION_EQUAL 15` for `-latomic`**

This change drops `-latomic` linking for Clang 16 and newer. I wasn't sure if that was intentional or a side-effect of the sanitizer work — could you share the motivation? If the goal was just to keep Clang 20 from picking up `-latomic`, a `VERSION_LESS 16` (or an explicit version range) would preserve the behavior for Clang 16/17/18/19 builds that may still need it.

**3. `SkewedPartitionBalancerTest.cpp` — disabled `serializedRebalanceExecution`**

The comment notes the test returns `2 vs 1 expected` or times out under `-O1`. That feels like it could be a real signal (optimizer interaction or a timing-sensitive test) rather than a sanitizer artifact. Would you mind opening a tracking issue and referencing it next to the `#ifndef`, the way the alignment/vptr suppressions reference #15810 / #15811? That way it doesn't get forgotten.

**4. `SignatureParser.yy` — the `yyFlexLexer` dummy block**

A few small things on this one:
- The dummies throw raw `std::runtime_error` — would `VELOX_FAIL` be a better fit for codebase consistency?
- The `\"- dummy\"` suffix on the `yylex()` message stood out a bit; happy to leave it but figured I'd flag it.
- More substantively: since the typeinfo is only missing under sanitizers, this looks more like a linker visibility / link-order issue than a sanitizer issue per se. Did you consider an explicit-instantiation TU or an \`extern template\` outside the \`#ifdef\`? That would localize the workaround and not depend on the sanitizer macro. Totally fine to defer if you've already explored that and ruled it out — just curious what you found.

**5. `ExchangeClientTest.cpp` — `executor_->join()` in `TearDown`**

This looks like a real lifetime fix that ASAN surfaced. Could you add a sentence in the commit message about what ASAN reported here? And do you happen to know whether any other test fixtures using `executor_` may need the same change, or is this one specific to the exchange-client lifecycle? Not asking you to audit them all — just want to make sure we're not patching one symptom of a broader pattern.

Thanks again — happy to chat on any of these if it's easier than going back and forth in comments.

czentgr commented:
@kgpai I've pushed some changes based on your review comments

1. linux-build-base.yml — VELOX_BUILD_SHARED=ON vs. the comment above it

The comment was outdated. I fixed it.

2. CMakeLists.txt — VERSION_GREATER_EQUAL 15 → VERSION_EQUAL 15 for -latomic

The implementation for the atomic symbols can come from the gcc atomic library (needing to add -latomic) or the `compiler-rt` library. The latter was added for asan/ubsan which is why we don't need to link atomic anymore.
To make things a bit more clear I added a function to check if compiler-rt is available instead of the version check.

3. SkewedPartitionBalancerTest.cpp — disabled serializedRebalanceExecution

This test is only run if enabled in DEBUG mode. But we build in RELEASE mode with `-O1`. As a result the test is not run.
I will check if the issue still occurs.
Edit: The issue does not occur anymore and no changes are made to the test.


4. SignatureParser.yy — the yyFlexLexer dummy block

I made it its own translation unit and added it to the common library because it appears multiple times across the code base.

5. ExchangeClientTest.cpp — executor_->join() in TearDown

It seems that this is a common pattern to ensure we don't have left over threads. I see others use a reset() function which is not implemented for this executor. I suspect adding the asan/ubsan symbols changed the timing a bit and so we need to ensure we are cleaning up. This is a test only problem.

kgpai commented:
Spotted an inconsistency between the CMake gate and the CI step:

The gate in `CMakeLists.txt` requires `CMAKE_BUILD_TYPE STREQUAL "Release"`:

```cmake
if(VELOX_ENABLE_ASAN_UBSAN_SANITIZERS
   AND CMAKE_BUILD_TYPE STREQUAL "Release"
   AND CMAKE_CXX_COMPILER_ID STREQUAL "Clang"
   AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 20)
```

But the workflow step `Make Asan/Ubsan Build (release plus -O1)` invokes `make debug`, which resolves to `BUILD_TYPE=Debug` in the Velox Makefile. So `CMAKE_BUILD_TYPE` ends up as `Debug` and the entire sanitizer block is skipped — no `-fsanitize=address`, no `-fsanitize=undefined`. The job ends up producing a vanilla Debug build under the ASAN/UBSAN job name.

Question: does the gate actually need to be Release-only? A Debug build with sanitizers would give better stack traces and line numbers in sanitizer reports, which is usually what you want when debugging the issues these tools flag.

czentgr committed:
feat(ci): Add ASAN and UBSAN build and CI job

czentgr commented:
@kgpai Thanks for the review. You were correct and it was a good catch. I missed to change this when switching from release back to debug. I addressed the issue. It should be good to go now.
We can see the sanitizer flags being set
```
FINAL CMAKE_CXX_FLAGS= -mavx2 -mfma -mavx -mf16c -mlzcnt -mbmi2 -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -Wall -Wextra -Wno-unused        -Wno-unused-parameter        -Wno-sign-compare        -Wno-ignored-qualifiers        -Wno-range-loop-analysis          -Wno-mismatched-tags          -Wno-nullability-completeness -Werror -O1 -fsanitize=address -fsanitize=undefined -fno-omit-frame-pointer -DVELOX_ENABLE_ASAN_UBSAN_SANITIZERS
```

kgpai commented:
Please fix the pre-commit issue, looks good otherwise.

czentgr commented:
@kgpai Yes, the pre-commit issue in unrelated. I've pinged you about it.

czentgr commented:
@kgpai CI is green now after rebasing again. Please take another look.

kgpai commented:
@czentgr Should land soon.
